{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b189babc",
   "metadata": {},
   "source": [
    "# Week 5: Simple single-cell RNA-seq analysis\n",
    "\n",
    "1. fetch some single-cell data and create a cell-gene expression matrix and the associated AnnData object;\n",
    "2. cluster the data to understand the cell structure; and\n",
    "3. annotate the data with biologically relevant information.\n",
    "\n",
    "All steps are based on Single-cell Best Practices Book (https://www.sc-best-practices.org/). Feel free to use it! You will still need to automate the process and ensure that it can be run."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d36a3341-7120-41ad-a56d-59b4e180b1b3",
   "metadata": {},
   "source": [
    "## Step 0: Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "294e7094",
   "metadata": {},
   "source": [
    "### Install Dependencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8f92a3ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/nick/miniconda3/envs/scrna/lib/python3.10/site-packages/pyranges/__init__.py:18: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  import pkg_resources\n"
     ]
    }
   ],
   "source": [
    "from pathlib import Path\n",
    "import urllib.request\n",
    "import subprocess\n",
    "import gzip\n",
    "import shutil\n",
    "import subprocess\n",
    "import pyroe\n",
    "import numpy as np\n",
    "import tarfile\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e513b54",
   "metadata": {},
   "source": [
    "### Set up working directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e2c05f0f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Working directory: /mnt/c/Users/Nick/documents/github/fall25-csc-bioinf/week6/code\n",
      "Data directory: /mnt/c/Users/Nick/documents/github/fall25-csc-bioinf/week6/code/data\n"
     ]
    }
   ],
   "source": [
    "notebook_dir = Path.cwd()\n",
    "data_dir = notebook_dir / \"data\"\n",
    "data_dir.mkdir(parents=True, exist_ok=True)\n",
    "print(f\"Working directory: {notebook_dir}\")\n",
    "print(f\"Data directory: {data_dir}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1eaa2819",
   "metadata": {},
   "source": [
    "## Step 1:\n",
    "\n",
    "Get the sample data here: https://app.box.com/s/lx2xownlrhz3us8496tyu9c4dgade814. This data contains the single-cell FASTQ files, as well as the reference genome (chr5) and the GTF file with transcript information.\n",
    "\n",
    "The list of whitelist barcodes is available [here](https://github.com/f0t1h/3M-february-2018/raw/refs/heads/master/3M-february-2018.txt.gz)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "download_whitelist",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Whitelist already exists: /mnt/c/Users/Nick/documents/github/fall25-csc-bioinf/week6/code/data/whitelist.txt\n"
     ]
    }
   ],
   "source": [
    "whitelist_url = \"https://raw.githubusercontent.com/f0t1h/3M-february-2018/master/3M-february-2018.txt.gz\"\n",
    "whitelist_gz = data_dir / \"3M-february-2018.txt.gz\"\n",
    "whitelist_path = data_dir / \"whitelist.txt\"\n",
    "\n",
    "if not whitelist_path.exists():\n",
    "    print(\"Downloading 10x whitelist...\")\n",
    "    urllib.request.urlretrieve(whitelist_url, whitelist_gz)\n",
    "    print(\"Extracting whitelist...\")\n",
    "    with gzip.open(whitelist_gz, 'rb') as f_in:\n",
    "        with open(whitelist_path, 'wb') as f_out:\n",
    "            shutil.copyfileobj(f_in, f_out)\n",
    "    whitelist_gz.unlink()\n",
    "    print(f\"Whitelist saved to: {whitelist_path}\")\n",
    "else:\n",
    "    print(f\"Whitelist already exists: {whitelist_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "setup_paths",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting dataset\n",
      "R1 reads: data/toy_ref_read/toy_read_fastq/selected_R1_reads.fastq\n",
      "R2 reads: data/toy_ref_read/toy_read_fastq/selected_R2_reads.fastq\n",
      "Genome FASTA: data/toy_ref_read/toy_human_ref/fasta/genome.fa\n",
      "GTF annotation: data/toy_ref_read/toy_human_ref/genes/genes.gtf\n",
      "Files successfully extracted\n"
     ]
    }
   ],
   "source": [
    "data_dir = Path(\"data\")\n",
    "data_dir.mkdir(parents=True, exist_ok=True)\n",
    "toy_dataset_gz = data_dir / \"toy_read_ref_set.tar.gz\"\n",
    "\n",
    "print(\"Extracting dataset\")\n",
    "\n",
    "with tarfile.open(toy_dataset_gz, 'r:gz') as tar:\n",
    "    tar.extractall(path=data_dir)\n",
    "    \n",
    "toy_ref_read = data_dir / \"toy_ref_read\"\n",
    "fastq_dir = toy_ref_read / \"toy_read_fastq\"\n",
    "ref_dir = toy_ref_read / \"toy_human_ref\"\n",
    "r1_fastq = fastq_dir / \"selected_R1_reads.fastq\"\n",
    "r2_fastq = fastq_dir / \"selected_R2_reads.fastq\"\n",
    "genome_fa = ref_dir / \"fasta\" / \"genome.fa\"\n",
    "genes_gtf = ref_dir / \"genes\" / \"genes.gtf\"\n",
    "\n",
    "files_ok = True\n",
    "\n",
    "for filepath, description in [\n",
    "    (r1_fastq, \"R1 reads\"),\n",
    "    (r2_fastq, \"R2 reads\"),\n",
    "    (genome_fa, \"Genome FASTA\"),\n",
    "    (genes_gtf, \"GTF annotation\")\n",
    "]:\n",
    "    exists = filepath.exists()\n",
    "    print(f\"{description}: {filepath}\")\n",
    "    if not exists:\n",
    "        files_ok = False\n",
    "\n",
    "if files_ok:\n",
    "    print(\"Files successfully extracted\")\n",
    "else:\n",
    "    print(\"Some files are missing\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1eaa2819_align",
   "metadata": {},
   "source": [
    "## Step 2:\n",
    "\n",
    "Use Alevin-fry to align and quantify this data to the reference genome. See [here](https://www.sc-best-practices.org/introduction/raw_data_processing.html#a-real-world-example) for the installation steps."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0b3b988",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "3d34fbed",
   "metadata": {},
   "source": [
    "## Step 3: \n",
    "\n",
    "Perform cell clustering (via Leiden modularity algorithm). Output the clustering plot."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "511cf8d9-6485-4a35-86fb-1958db374aae",
   "metadata": {},
   "source": [
    "## Step 4:\n",
    "\n",
    "Perform automatic cell annotation via CellTypist. Annotate the plot with the cell types."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "864a7283-34d0-4da0-984f-059b14311df7",
   "metadata": {},
   "source": [
    "## Time Estimate: 8 Hours"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
