{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b189babc",
   "metadata": {},
   "source": [
    "# Week 5: Simple single-cell RNA-seq analysis\n",
    "\n",
    "1. fetch some single-cell data and create a cell-gene expression matrix and the associated AnnData object;\n",
    "2. cluster the data to understand the cell structure; and\n",
    "3. annotate the data with biologically relevant information.\n",
    "\n",
    "All steps are based on Single-cell Best Practices Book (https://www.sc-best-practices.org/). Feel free to use it! You will still need to automate the process and ensure that it can be run."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d36a3341-7120-41ad-a56d-59b4e180b1b3",
   "metadata": {},
   "source": [
    "## Step 0: Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "294e7094",
   "metadata": {},
   "source": [
    "### Install Dependencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f92a3ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import urllib.request"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e513b54",
   "metadata": {},
   "source": [
    "### Set up working directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e2c05f0f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Working directory: /mnt/c/Users/Nick/Documents/GitHub/fall25-csc-bioinf/week5/code\n",
      "Data directory: /mnt/c/Users/Nick/Documents/GitHub/fall25-csc-bioinf/week5/code/data\n"
     ]
    }
   ],
   "source": [
    "notebook_dir = Path.cwd()\n",
    "data_dir = notebook_dir / \"data\"\n",
    "data_dir.mkdir(parents=True, exist_ok=True)\n",
    "print(f\"Working directory: {notebook_dir}\")\n",
    "print(f\"Data directory: {data_dir}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1eaa2819",
   "metadata": {},
   "source": [
    "## Step 1:\n",
    "\n",
    "Get the sample data here: https://app.box.com/s/lx2xownlrhz3us8496tyu9c4dgade814. This data contains the single-cell FASTQ files, as well as the reference genome (chr5) and the GTF file with transcript information.\n",
    "\n",
    "The list of whitelist barcodes is available here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60a145cd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DOWNLOADING REFERENCE GENOME\n",
      "Downloading chromosome 10 from UCSC Genome Browser\n",
      "Source: https://hgdownload.soe.ucsc.edu/goldenPath/hg38/chromosomes/chr10.fa.gz\n",
      "Target: /mnt/c/Users/Nick/Documents/GitHub/fall25-csc-bioinf/week5/code/data/chr10.fa.gz\n",
      "Download complete!\n",
      "Decompressing chr10.fa.gz...\n",
      "Decompression complete!\n",
      "Output file: /mnt/c/Users/Nick/Documents/GitHub/fall25-csc-bioinf/week5/code/data/chr10.fa\n",
      "Removed compressed file: /mnt/c/Users/Nick/Documents/GitHub/fall25-csc-bioinf/week5/code/data/chr10.fa.gz\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Download chromosome 10 reference genome\n",
    "toy_read_ref_set_url = \"https://app.box.com/s/lx2xownlrhz3us8496tyu9c4dgade814\"\n",
    "toy_read_ref_set_gz = data_dir / \"chr10.fa.gz\"\n",
    "\n",
    "# the gz file contains \n",
    "# toy_ref_read \n",
    "#   toy_read_fastq\n",
    "#      selected_R2_reads.fastq\n",
    "#      selected_R1_reads.fastq\n",
    "#   toy_read_human_ref\n",
    "#     genes\n",
    "#       genes.gtf\n",
    "#     fasta\n",
    "#       genome.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1eaa2819_align",
   "metadata": {},
   "source": [
    "## Step 2:\n",
    "\n",
    "Use Alevin-fry to align and quantify this data to the reference genome. See here for the installation steps."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51bd8635",
   "metadata": {},
   "source": [
    "## Step 3\n",
    "\n",
    "Use Alevin-fry to align and quantify this data to the reference genome. See here for the installation steps."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d34fbed",
   "metadata": {},
   "source": [
    "## Step 4: \n",
    "\n",
    "Perform cell clustering (via Leiden modularity algorithm). Output the clustering plot."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "511cf8d9-6485-4a35-86fb-1958db374aae",
   "metadata": {},
   "source": [
    "## Step 5:\n",
    "\n",
    "Perform automatic cell annotation via CellTypist. Annotate the plot with the cell types."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "864a7283-34d0-4da0-984f-059b14311df7",
   "metadata": {},
   "source": [
    "## Time Estimate: 8 Hours"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
